### ISSUE TITLE
Investigate lengthy data collection in python 3.10

### ISSUE THREAD

**Issue by mtrofin** [google/ml-compiler-opt#146]
All else being the same, compiling ~100 modules takes ~4-5s when using python3.9, but blows up to 30s when moving to python 3.10.

**Comment by mtrofin** [comment]
So far, using viztrace, was able to pinpoint protobuf serialization as very likely root cause. The protobuf package is at the same version, though, so looking further.

**Comment by mtrofin** [comment]
It seems that updating to `protobuf==3.19.5` addresses that. Would prefer postpone changing the version in `requirements.txt` to when we bump the tf version, and this is another reason to move away from protobuf altogether.

**Comment by boomanaiden154** [comment]
I believe the protobuf version has been updated to `3.19.5` in the version of Tensorflow that we're already using (or the version is just set to be `> 3.x.x && < 4.0.0`. When I regenerated the lockfile in #128, it automatically grabbed `3.19.5` as when we upgraded to Tensorflow nightly originally, we didn't really touch any of the dependency versions (#118).

**Comment by mtrofin** [comment]
huh. so I must have quite a messed up local setup.

**Comment by boomanaiden154** [comment]
It's not your local setup. The protobuf version in `requirements.txt` is still set to the lower version.

**Comment by boomanaiden154** [comment]
Testing after #189:
Default trace over an LLVM corpus, default settings (Python 3.8.10):
```
real	1m43.543s
user	120m39.134s
sys	4m22.878s
```
Warmstart:
```
real	1m32.171s
user	4m54.158s
sys	3m45.621s
```
PPO training, default settings (except setting `num_policy_iterations`):
```
real	6m42.992s
user	219m37.325s
sys	40m12.776s
```
Default trace over an LLVM corpus, default settings (Python 3.10.6):
```
real	1m41.111s
user	121m11.636s
sys	2m33.097s
```
Warmstart:
```
real	1m30.252s
user	5m6.882s
sys	3m32.267s
```
PPO training, default settings (except setting `num_policy_iterations`):
```
real	6m32.612s
user	199m24.214s
sys	34m36.932s
```
While these runs aren't perfectly controlled to just the data collection step (although the default trace runs should come close), they do show that the performance is pretty comparable between 3.10 and previous versions now. I'm seeing similar times for module compilation between both versions and about the performance I'd "expect" given timings/performance that I've seen before.

I compiled some documentation on what I did in this [gist](https://gist.github.com/boomanaiden154/ab014fcf1b57b943e3bb19fe78d9ebcd). Given that there doesn't seem to be performance differences between the versions now, I'm going to close this issue.
